frontend/src/pages/RequestorPage/ResponsePanel.tsx

```diff
@@ -545,31 +545,118 @@ function ResponseBodyBinary({
 
 export function ResponseBodyText({
   body,
+  maxPreviewLength = null,
+  maxPreviewLines = null,
+  defaultExpanded = false,
   className,
-}: { body: string; className?: string }) {
+}: {
+  body: string;
+  maxPreviewLength?: number | null;
+  maxPreviewLines?: number | null;
+  defaultExpanded?: boolean;
+  className?: string;
+}) {
+  const [isExpanded, setIsExpanded] = useState(!!defaultExpanded);
+  const toggleIsExpanded = () => setIsExpanded((e) => !e);
+
   // For text responses, just split into lines and render with rudimentary line numbers
-  const lines = useMemo(() => {
-    return body?.split("\n")?.map((line, index) => (
-      <div key={index} className="flex h-full">
-        <span className="w-8 text-right pr-2 text-gray-500 bg-muted mr-1">
-          {index + 1}
-        </span>
-        <span>{line}</span>
-      </div>
-    ));
-  }, [body]);
+  const { lines, hiddenLinesCount, hiddenCharsCount, shouldShowExpandButton } =
+    useTextPreview(body, isExpanded, maxPreviewLength, maxPreviewLines);
 
   // TODO - if response is empty, show that in a ux friendly way, with 204 for example
 
   return (
-    <div className={cn("overflow-hidden overflow-y-auto w-full", className)}>
+    <div
+      className={cn("overflow-hidden overflow-y-auto w-full py-2", className)}
+    >
       <pre className="text-sm font-mono text-gray-300 whitespace-pre-wrap">
         <code className="h-full">{lines}</code>
       </pre>
+      {shouldShowExpandButton && (
+        <div
+          className={cn(
+            "w-full flex flex-row items-center gap-2 mt-4 border-t border-gray-500/50",
+          )}
+        >
+          {!isExpanded && (
+            <div className="text-sm text-gray-400">
+              {hiddenLinesCount > 0 ? (
+                <>{hiddenLinesCount} lines hidden</>
+              ) : (
+                <>{hiddenCharsCount} characters hidden</>
+              )}
+            </div>
+          )}
+          <Button
+            variant="ghost"
+            onClick={toggleIsExpanded}
+            className="text-blue-400 font-normal py-1 px-1 hover:bg-transparent hover:text-blue-400 hover:underline"
+          >
+            {isExpanded ? "Show Less" : "Show More"}
+          </Button>
+        </div>
+      )}
     </div>
   );
 }
 
+function useTextPreview(
```

Comment:
Doesn't need to be part of this pr, but this file is really ripe to be split up into separate parts: just make a directory

The ResponseBodyText component for instance could benefit from its own file especially since it's been exported.

Reply:
did it!